### ISSUE TITLE
Bad parsing recovery on missing for with impl

### ISSUE THREAD

**Issue by bjorn3** [rust-analyzer/rust-analyzer#921]
rust-analyzer gives errors at the specified locations:

```rust
impl<'tcx> rustc::ty::fold::TypeVisitor<'tcx> UI128Visitor {
//                                           ^
    fn visit_ty(&mut self, t: Ty<'tcx>) -> bool {
//  ^ ^        ^^^  ^^           ^    ^^^ ^    ^^
        t.super_visit_with(self)
    }
}
```

Rustc gives just one error and doesn't abort parsing (proof: adding syntax error in fn body is catched):

```
error: missing `for` in a trait impl
  --> src/base.rs:75:54
   |
75 | impl<'tcx> rustc::ty::fold::TypeVisitor<'tcx> UI128Visitor {
   |                                              ^ help: add `for` here
```

**Comment by DJMcNab** [comment]
I don't think we hardcode any better errors for specific mistakes, but it is always worth adding specific recovery paths. [I'd say this is help-wanted E-easy, but only P-low (if we had such a designation)]. 

To implement this, if at the impl item parsing, when we check for the `FOR` keyword we should check if we are at_ts `PATH_START`, then parse a path and try and parse the inner items then

**Comment by matklad** [comment]
Filed https://github.com/rust-analyzer/rust-analyzer/pull/925 which should avoid cascading erros in similar cases. #925 will parse the contents of any unrecognized block as an expression. 

Adding a specific case for missing `for` will be useful, but I don't think this would have a good cost/benefit ratio at this stage: we are not yet aiming at producing good error messages.
